Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp

```diff
@@ -67,6 +67,79 @@ struct TextureMapperLayer::ComputeTransformData {
     }
 };
 
+class TextureMapperFlattenedLayer final {
+    WTF_MAKE_FAST_ALLOCATED;
+public:
+    explicit TextureMapperFlattenedLayer(const IntRect& rect, double zNear, double zFar)
+        : m_rect(rect)
+        , m_zNear(zNear)
+        , m_zFar(zFar)
+    { }
+
+    inline IntRect layerRect() const
+    {
+        return m_rect;
+    }
+
+    inline bool valid() const
+    {
+        return m_textures.size() > 0;
+    }
+
+    void update(TextureMapperPaintOptions& options, const std::function<void(TextureMapperPaintOptions&)>& paintFunction)
+    {
+        double prevZNear, prevZFar;
+        options.textureMapper.getDepthRange(prevZNear, prevZFar);
+        options.textureMapper.setDepthRange(m_zNear, m_zFar);
+
+        IntSize maxTextureSize = options.textureMapper.maxTextureSize();
+        for (int x = m_rect.x(); x < m_rect.maxX(); x += maxTextureSize.width()) {
+            for (int y = m_rect.y(); y < m_rect.maxY(); y += maxTextureSize.height()) {
+                IntRect tileRect({ x, y }, maxTextureSize);
+                tileRect.intersect(m_rect);
+
+                auto surface = options.textureMapper.acquireTextureFromPool(tileRect.size(), { BitmapTexture::Flags::SupportsAlpha });
+                {
+                    SetForScope scopedSurface(options.surface, surface.ptr());
+                    SetForScope scopedOffset(options.offset, -toIntSize(tileRect.location()));
+                    SetForScope scopedOpacity(options.opacity, 1);
+
+                    options.textureMapper.bindSurface(options.surface.get());
+                    paintFunction(options);
+
+                    surface = Ref { *options.surface };
```

Comment:
Why do you need this line?

Reply:
It's for the case when the element which initiates the flattening layer has a filter for example drop shadow. When TextureMapper applies the effect it replaces the original surface and sets new surface to options.surface. The way how filters work for flattened layer is that the filter is applied to whole layer. Thus we need to render this new surface texture that filter assigned instead the original flattened surface texture.
I added a text case to bug https://bug-281079-attachments.webkit.org/attachment.cgi?id=473059

Reply:
I got it. paintIntoSurface create another BitmapTexture. Never mind.